### PR TITLE
Add dynamic tiling experience for floating windows.

### PR DESCRIPTION
**Pull request by jcguu95**
# Features

(c.f. #881)

+ Rotate
+ Permute
+ Float, move, resize, unfloat
+ Float all, unfloat all.
+ Toggle fullscreen layout

### DEMO

+ Basic Movement

Basic Movement

+ Float and Retile

Float and Retile

+ Layout and Ratio

Layout and Ratio

# Commit Message

* AUTHORS

  Add author: Jin-Cheng Guu.

* dynamic-floating-group.lisp

  Main body of new codes that provides dynamic tiling experience for
  floating windows.

* package.lisp

  All new symbols are created under a new CL package `:stumpwm-dfg`.

* stumpwm.asd

  Add `dynamic-floating-group.lisp` as a component of the system.

### CODE REVIEW

**Comment by jcguu95** [comment]
More documentations to come after this is merged. In the meanwhile, please test it out and provide feedback. I'm willing to patch any issues if any. Discussions could help the pull request to be merged, I believe.

**Comment by jcguu95** [comment]
My current key-bindings for it to work is included in this note.

Note however that the function names might change after it's merged. I do expect to change some of their names, so what I said below won't work out of the box in the future. I do however plan to write a documentation for it after it's merged. For now, let's stick with the following.

---

Almost all commands I introduced worked exclusively for dynamic-floating-groups. So first we must make a new dynamic-floating-group.

```
(stumpwm-dfg::gnew-dyn-float "test")
```

It should bring you to that freshly-created group immediately. Now start to open some windows. They will be floating windows, and should be tiled automatically within that group.

The following is my binding at the moment. It has been working for a few weeks. Notice that the `wrappers` are just a dirty/temporary hack to use the same keybinding for different groups. However, as I have lived in my dynamic-floating-group for a while without any problem, I will move completely into it, making the dirty hacks obsolete!

```
(defcommand tmp-wrapper-s-j () ()
  (let ((cg (stumpwm:current-group)))
    (if (dyn-float-group-p cg)
        (focus-next-window)
        (call-interactively 'move-focus "down"))))
(define-key *top-map* (stumpwm:kbd "s-j") "tmp-wrapper-s-j")

(defcommand tmp-wrapper-s-k () ()
  (let ((cg (stumpwm:current-group)))
    (if (dyn-float-group-p cg)
        (focus-last-window)
        (call-interactively 'move-focus "up"))))
(define-key *top-map* (stumpwm:kbd "s-k") "tmp-wrapper-s-k")

(defcommand tmp-wrapper-s-capitol-j () ()
  (let ((cg (stumpwm:current-group)))
    (if (dyn-float-group-p cg)
        (permute-window-list)
        (call-interactively 'exchange-direction "down"))))
(define-key *top-map* (stumpwm:kbd "s-J") "tmp-wrapper-s-capitol-j")

(defcommand tmp-wrapper-s-capitol-k () ()
  (let ((cg (stumpwm:current-group)))
    (if (dyn-float-group-p cg)
        (permute-window-list t)
        (call-interactively 'exchange-direction "up"))))
(define-key *top-map* (stumpwm:kbd "s-K") "tmp-wrapper-s-capitol-k")
(define-key *top-map* (stumpwm:kbd "s-_") "unfree-all")
(define-key *top-map* (stumpwm:kbd "s-+") "increase-master-ratio")
(define-key *top-map* (stumpwm:kbd "s--") "decrease-master-ratio")
(define-key *top-map* (stumpwm:kbd "s-=") "default-master-ratio")
(define-key *top-map* (stumpwm:kbd "s-f") "toggle-fullscreen-layout")

```

**Review by PuercoPop — COMMENTED** [review]
Sorry, I haven't really finished with the review. But I didn't want to leave you hanging for so long. So I'm submitting a partial review. Before merging to the main StumpWM this also needs to be documented on the manual.

As you mentioned on reddit, if you where to submit this as a contrib the requirements as far lower before merging, mostly it needs to build and have a short readme so if you prefer to take that route you can submit this as a contrib. If you do so I'm fine merging the code as is.

**Comment by jcguu95** [comment]
Wooosh this is a lot of effort made! Thank you @PuercoPop! I will fix them and add documentations once you are done. 

Given the effort.. let's put it into the main repo then! I will do my best to make it as good as possible :) Thanks again!

**Comment by dmb2** [comment]
As soon as this is documented and the code review complete we can merge.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:21]: Added to the docstring of `*default-layout*`.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:57]: I'm not quite sure what you mean.. would you please elaborate a bit more?

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:113]: Resolved as you suggested.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:1]: Added the year.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:19]: To do this I should rename some symbols like `*default-layouts*` in the package `stumpwm-dfg`. I will probably do this after other things are done.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: Nice foo! I have resolved every such thing by `cl:assert`.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: Resolved as suggested.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: Resolved as suggested.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:32]: ### No keywords for slots.

Ah! My bad. I have resolved this accordingly.

### sub-class

What do you mean? I am new to CLOS and might not use the best practice. So far, I've learned to create a sub-class of an existing class. But I don't know how to create a super-class of an existing class, making the latter a sub-class. Is this what you mean?

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: I've replaced `nconc` by `alexandria:appendf`.

However, I think the later still modifies the list destructively. But it's better as it does not seem to create cyclic (infinite) lists.

File: dynamic-floating-group.lisp
Comment (PuercoPop) [reply dynamic-floating-group.lisp]: So the code currently defines a struct window+ that wraps a window and adds a slot/attribute, free. I'm suggesting that we define a subclass of window instead with that extra slot instead. Something like

```lisp
(defclass window+ (window)
  ((free :initarg :free :reader window+-free)))
```

We can use change-class to update the class of the window to the appropriate sub-class like we do for floating windows. What do you think about this suggestion?

File: dynamic-floating-group.lisp
Comment (PuercoPop) [reply dynamic-floating-group.lisp:60]: I'm saying that instead of repeating the same piece of code in two different local function definitions, lets define a top level function and use it on both places

File: dynamic-floating-group.lisp
Comment (PuercoPop) [reply dynamic-floating-group.lisp:19]: We can keep the code in a separate package then. What do you think about importin the internal functions we are using from the stumpwm package to avoid the `stumpwm::` clutter?

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: Make sense, and done - [6a318f3](https://github.com/stumpwm/stumpwm/pull/885/commits/6a318f31cd3a6db72963073c8d764546c6023b65).

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: Got it! Done in [7e0fee4](https://github.com/stumpwm/stumpwm/pull/885/commits/7e0fee400e7c301b08371905ff65b0aceb6b2877).

+ Please note that this is the first time this branch changes the old code!
+ Please also note that `stumpwm:` will be taken away from `stumpwm:add-float-window` later.

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp:32]: Are you sure? I am happy to leave it in a separate package too. Perhaps a few years later when people find it stable, we can merge them back anyways. It shouldn't be hard.

About the `stumpwm::` clutter: I think it's ok to have them there. I have a few `stumpwm:` anyways, and having an extra `:` doesn't look like much more clutter to me. 

Let's leave it as is, wait for a while, and decide then if merging is needed? At this moment, leaving them separated makes it easier to debug IMO. Though I have been using only it for the past 2~3 months without issues. Please let me know!

Comment (jcguu95) [reply dynamic-floating-group.lisp:32]: I decide to leave it as is. Let's consider merging the package later in the future.

**Comment by jcguu95** [comment]
+ @PuercoPop all of your suggestions have been addressed. 
+ @dmb2  A temporary manual has been added in the latest commit. If you like how it is, please advise me how to merge it into the official one.

Thanks!

**Comment by PuercoPop** [comment]
@jcguu95 Ideally we would use change-change to change a window to a window+. But a decorator pattern is ok as well. I've been trying to find the time to try to change the code myself but I've been unable. It seems unlikely that I'll be able to do so in the near future. The code is already vastly improved. I'm OK with us merging the code with the current decorator approach. The only thing I'll ask is that if possible to add some comments at the top of the dynamic-floating-group files explaining the overall strategy used (especially the ratios).

@jcguu95 Once again, thank you for your work _and_ thanks for your patience!

File: dynamic-floating-group.lisp
Comment (PuercoPop) [inline dynamic-floating-group.lisp:147]: We can use sort for this. See the following snippet

```lisp
* (sort (list t nil t nil t) (lambda (x y) x))
(T T T NIL NIL)
```

So something like
```lisp
;; We define a predicate as a top level form outside this function
(defun drifting-first (win-a win-b)
  (declare (ignore win-b))
  (window+-drift win-a))
...  
    ;; Make the drifting windows on top of the stack.
    (setf dyn-order
      (sort (copy-list dyn-order) #'drifting-first))
```

should work. A common CL gotcha is that sort destructively modifies the list which is why I'm adding the copy-list.

P.D.

The current approach could be made shorter by removing the surrounding (equal ..) form and using remove-if-not for the first filter. something like:

```lisp
    ;; Make the drifting windows on top of the stack.
    (setf dyn-order
          (append (remove-if-not #'window+-drift dyn-order)
                  (remove-if #'window+-drift dyn-order)))
```

But I think sort captures the intent of the code better.

**Review by PuercoPop — APPROVED** [review]

File: dynamic-floating-group.lisp
Comment (jcguu95) [reply dynamic-floating-group.lisp]: Nice fu! Done.

**Comment by jcguu95** [comment]
@PuercoPop 

1. Thank you for your patience and guidance too! I've learned a lot, and I am very glad that I can contribute :smile: 
2. Lets merge it if you are happy with that. I can see your approval. What else should I do?
3. What is the change-change method you mentioned?

**Comment by PuercoPop** [comment]
> Lets merge it if you are happy with that. I can see your approval. What else should I do?

Nothing, we are waiting for @dmb2's approval of the manual

> What is the change-change method you mentioned?

Sorry I mean change-class. That is converting the window instance into a window+ instead of 'wrapping' the instance by keeping it in a slot.

**Comment by jcguu95** [comment]
@PuercoPop It has been a while. I've implemented some more features in [#908](https://github.com/stumpwm/stumpwm/pull/908). Not sure if you want to check the code of that, and merge with this PR.

**Comment by jcguu95** [comment]
Is this still of interest?

**Comment by dmb2** [comment]
What's going with the README? I don't mind having separate docs for this, but it looks like this PR clobbers the default readme which I'm mildly fond of :).

**Comment by jcguu95** [comment]
Ah I made the unofficial branch easier for people who just want to have a try. Forgot that it affects this PR. The last commit that's suitable for this PR is "d825304b01eb5530b115299f46cf2657e9743e5b". When it's ready to merge, I will change it back. Please let me know!

**Comment by szos** [comment]
I have a small concern about this, namely that I think its best for StumpWM to have one canonical dynamic tiling group implementation. Personally, I think that if this is merged then the current dynamic tiling group implementation should be removed. I'm curious what others think about this? Is it best to have one implementation, or should there be multiple? I think it could be confusing to newcomers to have two group types that do effectively the same thing.

**Comment by jcguu95** [comment]
I'm happy to turn this into a contrib.

**Comment by szos** [comment]
While that would be the simplist option/path of least resistance, i think it would be worth getting opinions from the maintainers on which solution they prefer before pulling any triggers. After all, if we're hit by a buss theyre the ones who will need to maintain it. 

Im currently investigating adding support for multiple heads and realized that when/if your solution is merged we're gonna be working twice as much for the same number of solutions to any future problems, hence my comment on having one canonical implementation. 

Id be lying if i said i didnt prefer my own implementation, though thats mostly due to understanding how it works on a deeper level than i understand your implementation. Because of that bias i hesitate to make definitive statements on which should be the canonical implementation, and prefer to leave that decision up to the maintainers. 

All that being said, the pros and cons as i see them of making this into a contrib are: 

Pros: merging will be easier. Making radical changes will be easier. More personal control over direction and behavior of the implementation.

Cons: breakages may be more common, as folks tend not to test against the contrib repo when making changes (this is dependent upon how ingrained your code is with stumpwm internals).

There are likely other pros/cons of making this a contrib that im not thinking of.

**Comment by dmb2** [comment]
I would like one implementation. Naively, it seems like this can be accomplished by floating/unfloating windows using the existing dynamic windows code.

**Comment by szos** [comment]
My apologies, but I'm a touch confused - what does the word «this» refer to in the sentence «[...] it seems like this can be accomplished by floating/unfloating windows [...]»?

**Comment by dmb2** [comment]
'this' being the topic of the pull request: dynamically tiling floating windows.  Maybe I'm not following the proposed feature that's being added.  We have dynamic tiling already.  From the demo graphics it looks like this does the same but allows you to float windows on top.

**Comment by szos** [comment]
Thanks for clarifying. Based on my understanding of this PR (from the videos and a brief glance at the code), this is another implementation of dynamic tiling using floating windows instead of tiling windows. Full disclosure, I havent tested this PR at all, so take what I say with a grain of salt. The current dynamic tiling implementation¹ allows for floating windows, which are then managed purely by the user. This implementation uses only floating windows (wrapped in another class) which does (to my knowledge) the same thing as the current implementation but in a different fashion. 

¹I'm currently rewriting the current dynamic tiling implementation to implement support for multiple heads and clean up the code (add comments, etc).

**Comment by jcguu95** [comment]
I started this project to ease more people like me, who are used to a floating settings, to enter the wonderful realm of StumpWM. However, I also understand that the purpose of this extension could be confusing to the new users. So at this point, it seems that the best move is to close this PR. I will try to make my extension as a module for people who are interested. Thank you all so much for all thoughtful discussion. Special thanks to @PuercoPop for code review. Please feel free to close this PR.

**Comment by Kreyren** [comment]
Provide an installation method assuming it being finished then i am stalking this MR daily for like 2 months to get floating in stumpWM

**Comment by jcguu95** [comment]
@Kreyren I rewrote the whole thing [recently](https://github.com/jcguu95/stumpwm-dynamic-float). Now it has less lines of codes, better behavior, and better customizability. There, you can find an installation method. Feel free to open an issue if you need any help. Good luck, and thanks for asking.

**Comment by dmb2** [comment]
Apologies to @jcguu95, @szos has precedent because his dynamic code was merged to master earlier. He re-wrote it for multiple heads.  Does the simplifications of the refactor allow you to accomplish what you want using his code?

**Comment by jcguu95** [comment]
No worries! I thank you a lot for all the effort :)

I haven't checked his new code. Does it allow us to float windows, drag windows, and resize windows?

**Comment by szos** [comment]
The old and new implementation both support floating windows either via float this or the :float argument to define frame preference. The new implementation allows for resizing the master/stack split, but continuously balances the stack frames. Im unsure what window dragging functionality you would like, but it should be easy-ish to implement a method for it (if i remember correctly, the floating window resize/move functionality is a method dispatching on floating windows). If you specify the drag functionality i can try to figure out how to implement it.

--
Szos

12. jan. 2022 13:10:04 Jin ***@***.***>:

**Pull request closed without merging**
